<a href="https://colab.research.google.com/github/e23046/Statistical-Learning-e23046/blob/main/Assignment%2003/Q1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Problem: Fraud Detection

A bank monitors credit-card transactions for possible fraud. We are given a sample space $\Omega$, probabilities for each outcome, and two random variables $X$ (fraud indicator) and $Y$ (customer risk class).

Let's first define the sample space, probabilities, and the random variables for calculations.

In [2]:
import numpy as np
import pandas as pd

# Define the sample space outcomes and their probabilities
omega = {'omega_1': 0.50, 'omega_2': 0.10, 'omega_3': 0.20, 'omega_4': 0.20}
p_omega = np.array(list(omega.values()))

# Define the random variable X (fraud indicator)
X_values = {'omega_1': 0, 'omega_2': 1, 'omega_3': 0, 'omega_4': 1}

# Define the random variable Y (customer risk class)
Y_values = {'omega_1': 0, 'omega_2': 0, 'omega_3': 1, 'omega_4': 1}

print("Probabilities P(omega_i):")
for k, v in omega.items():
    print(f"{k}: {v}")

print("\nX values:")
for k, v in X_values.items():
    print(f"{k}: {v}")

print("\nY values:")
for k, v in Y_values.items():
    print(f"{k}: {v}")

Probabilities P(omega_i):
omega_1: 0.5
omega_2: 0.1
omega_3: 0.2
omega_4: 0.2

X values:
omega_1: 0
omega_2: 1
omega_3: 0
omega_4: 1

Y values:
omega_1: 0
omega_2: 0
omega_3: 1
omega_4: 1


### 1. Probability Space $(\Omega,\mathfrak F,P)$

The probability space is defined as $(\Omega,\mathfrak F,P)$, where:

*   **Sample Space $\Omega$**: The set of all possible outcomes.
    $$ \Omega=\{\omega_1,\omega_2,\omega_3,\omega_4\} $$

*   **Event Space $\mathfrak F$**: The power set of $\Omega$, denoted by $\mathfrak P(\Omega)$. This includes all possible subsets of $\Omega$. Since $\Omega$ has 4 elements, $\mathfrak P(\Omega)$ has $2^4 = 16$ elements. These include the empty set $\emptyset$, each individual $\omega_i$, all pairs, all triples, and $\Omega$ itself.

*   **Probability Measure $P$**: A function that assigns a probability to each event in $\mathfrak F$. For individual outcomes, the probabilities are given as:
    $$ P(\{\omega_1\})=0.50,\qquad P(\{\omega_2\})=0.10,\qquad P(\{\omega_3\})=0.20,\qquad P(\{\omega_4\})=0.20. $$
    For any other event $A \in \mathfrak F$, $P(A)$ is the sum of the probabilities of the individual outcomes in $A$. For example, $P(\{\omega_1, \omega_2\}) = P(\{\omega_1\}) + P(\{\omega_2\}) = 0.50 + 0.10 = 0.60$.

### 2. Compute the $\sigma$-algebra generated by $Y$, namely $\sigma(Y)$.

The $\sigma$-algebra generated by a random variable $Y$, denoted $\sigma(Y)$, is the smallest $\sigma$-algebra with respect to which $Y$ is measurable. It is formed by taking all preimages of measurable sets in the codomain of $Y$.

First, let's look at the values $Y$ takes and their corresponding preimages:
*   $Y=0$ when $\omega = \omega_1$ or $\omega_2$. So, $Y^{-1}(\{0\}) = \{\omega_1, \omega_2\}$.
*   $Y=1$ when $\omega = \omega_3$ or $\omega_4$. So, $Y^{-1}(\{1\}) = \{\omega_3, \omega_4\}$.

The $\sigma$-algebra generated by $Y$ will consist of the empty set, $\Omega$, and all unions, intersections, and complements of these preimages.

$$ \sigma(Y) = \{\emptyset, \{\omega_1, \omega_2\}, \{\omega_3, \omega_4\}, \Omega \} $$

### 3. Explain why $E[X\mid Y]$ must be $\sigma(Y)$-measurable, and therefore must be constant on the sets $\{\omega_1,\omega_2\}$ and $\{\omega_3,\omega_4\}$.

By definition, the conditional expectation $E[X\mid Y]$ is a random variable that is $\sigma(Y)$-measurable. This means that for any measurable set $B$ in the range of $E[X\mid Y]$, its preimage $(E[X\mid Y])^{-1}(B)$ must belong to $\sigma(Y)$.

A key property of a random variable being $\sigma(Y)$-measurable is that it must be constant on the atoms of $\sigma(Y)$. The atoms of $\sigma(Y)$ are the smallest non-empty sets in $\sigma(Y)$ that cannot be further decomposed. In this case, the atoms of $\sigma(Y)$ are precisely the sets where $Y$ takes a constant value:

*   The set where $Y=0$: $\{\omega_1, \omega_2\}$
*   The set where $Y=1$: $\{\omega_3, \omega_4\}$

Since $E[X\mid Y]$ is $\sigma(Y)$-measurable, it must take a constant value on $\{\omega_1, \omega_2\}$ (where $Y=0$) and another constant value on $\{\omega_3, \omega_4\}$ (where $Y=1$). This makes intuitive sense: if we know $Y$, then our best estimate of $X$ should only depend on the value of $Y$, not on which specific $\omega_i$ occurred within the set where $Y$ is constant.

### 4. Compute $E[X\mid Y]$ explicitly and show that $E[X\mid Y]= \frac{1}{6} \mathbb{1}_{\{Y=0\}} + \frac{1}{2} \mathbb{1}_{\{Y=1\}}$

The conditional expectation $E[X \mid Y=y]$ is calculated as follows:

*   For $Y=0$ (low-risk customer), the relevant outcomes are $\{\omega_1, \omega_2\}$.
    The probability of this event is $P(Y=0) = P(\{\omega_1\}) + P(\{\omega_2\}) = 0.50 + 0.10 = 0.60$.

    $$ E[X \mid Y=0] = \frac{E[X \cdot \mathbb{1}_{\{Y=0\}}]}{P(Y=0)} = \frac{X(\omega_1)P(\{\omega_1\}) + X(\omega_2)P(\{\omega_2\})}{P(\{\omega_1\}) + P(\{\omega_2\})} $$
    $$ E[X \mid Y=0] = \frac{(0)(0.50) + (1)(0.10)}{0.50 + 0.10} = \frac{0.10}{0.60} = \frac{1}{6} $$

*   For $Y=1$ (high-risk customer), the relevant outcomes are $\{\omega_3, \omega_4\}$.
    The probability of this event is $P(Y=1) = P(\{\omega_3\}) + P(\{\omega_4\}) = 0.20 + 0.20 = 0.40$.

    $$ E[X \mid Y=1] = \frac{E[X \cdot \mathbb{1}_{\{Y=1\}}]}{P(Y=1)} = \frac{X(\omega_3)P(\{\omega_3\}) + X(\omega_4)P(\{\omega_4\})}{P(\{\omega_3\}) + P(\{\omega_4\})} $$
    $$ E[X \mid Y=1] = \frac{(0)(0.20) + (1)(0.20)}{0.20 + 0.20} = \frac{0.20}{0.40} = \frac{1}{2} $$

Therefore, the random variable $E[X\mid Y]$ can be written as:
$$ E[X\mid Y]= \frac{1}{6} \mathbb{1}_{\{Y=0\}} + \frac{1}{2} \mathbb{1}_{\{Y=1\}} $$

In [3]:
# Calculate E[X|Y=0]
P_Y0 = omega['omega_1'] + omega['omega_2']
EX_given_Y0 = (X_values['omega_1'] * omega['omega_1'] + X_values['omega_2'] * omega['omega_2']) / P_Y0
print(f"E[X|Y=0] = {EX_given_Y0:.4f} (which is 1/6)")

# Calculate E[X|Y=1]
P_Y1 = omega['omega_3'] + omega['omega_4']
EX_given_Y1 = (X_values['omega_3'] * omega['omega_3'] + X_values['omega_4'] * omega['omega_4']) / P_Y1
print(f"E[X|Y=1] = {EX_given_Y1:.4f} (which is 1/2)")

# Represent E[X|Y] as a random variable over omega
EX_given_Y_rv = {
    'omega_1': EX_given_Y0,
    'omega_2': EX_given_Y0,
    'omega_3': EX_given_Y1,
    'omega_4': EX_given_Y1
}

print("\nE[X|Y] as a random variable (values for each omega_i):")
for k, v in EX_given_Y_rv.items():
    print(f"{k}: {v:.4f}")

E[X|Y=0] = 0.1667 (which is 1/6)
E[X|Y=1] = 0.5000 (which is 1/2)

E[X|Y] as a random variable (values for each omega_i):
omega_1: 0.1667
omega_2: 0.1667
omega_3: 0.5000
omega_4: 0.5000


### 5. Verify the defining property of conditional expectation by checking that for every set $A\in\sigma(Y)$, $$ \int_A E[X\mid Y]\,dP=\int_A X\,dP. $$

The defining property of conditional expectation states that for any $A \in \sigma(Y)$:
$$ E[E[X|Y] \mathbb{1}_A] = E[X \mathbb{1}_A] $$
This is equivalent to the integral form. We need to check this for all sets in $\sigma(Y)$, which are $\{\emptyset, \{\omega_1, \omega_2\}, \{\omega_3, \omega_4\}, \Omega \}$.

1.  **For $A = \emptyset$**: Both sides are 0. $0=0$.

2.  **For $A = \{\omega_1, \omega_2\}$ (where $Y=0$)**:
    *   LHS: $E[E[X\mid Y] \cdot \mathbb{1}_{\{\omega_1, \omega_2\}}]$
        $$ = E[X\mid Y=0] P(\{\omega_1\}) + E[X\mid Y=0] P(\{\omega_2\}) $$
        $$ = \frac{1}{6} (0.50) + \frac{1}{6} (0.10) = \frac{1}{6} (0.60) = 0.10 $$
    *   RHS: $E[X \cdot \mathbb{1}_{\{\omega_1, \omega_2\}}]$
        $$ = X(\omega_1)P(\{\omega_1\}) + X(\omega_2)P(\{\omega_2\}) $$
        $$ = (0)(0.50) + (1)(0.10) = 0.10 $$
    LHS = RHS, so the property holds for this set.

3.  **For $A = \{\omega_3, \omega_4\}$ (where $Y=1$)**:
    *   LHS: $E[E[X\mid Y] \cdot \mathbb{1}_{\{\omega_3, \omega_4\}}]$
        $$ = E[X\mid Y=1] P(\{\omega_3\}) + E[X\mid Y=1] P(\{\omega_4\}) $$
        $$ = \frac{1}{2} (0.20) + \frac{1}{2} (0.20) = \frac{1}{2} (0.40) = 0.20 $$
    *   RHS: $E[X \cdot \mathbb{1}_{\{\omega_3, \omega_4\}}]$
        $$ = X(\omega_3)P(\{\omega_3\}) + X(\omega_4)P(\{\omega_4\}) $$
        $$ = (0)(0.20) + (1)(0.20) = 0.20 $$
    LHS = RHS, so the property holds for this set.

4.  **For $A = \Omega$**:
    *   LHS: $E[E[X\mid Y]]$
        $$ = E[X\mid Y=0] P(Y=0) + E[X\mid Y=1] P(Y=1) $$
        $$ = \frac{1}{6} (0.60) + \frac{1}{2} (0.40) = 0.10 + 0.20 = 0.30 $$
    *   RHS: $E[X] = \sum_{i=1}^4 X(\omega_i)P(\{\omega_i\})$
        $$ = (0)(0.50) + (1)(0.10) + (0)(0.20) + (1)(0.20) $$
        $$ = 0.10 + 0.20 = 0.30 $$
    LHS = RHS, so the property holds for $\Omega$.

Since the property holds for all sets in $\sigma(Y)$, it is verified.

In [4]:
# Verify for A = {omega_1, omega_2} (Y=0)
# LHS
LHS_A_Y0 = EX_given_Y0 * omega['omega_1'] + EX_given_Y0 * omega['omega_2']
# RHS
RHS_A_Y0 = X_values['omega_1'] * omega['omega_1'] + X_values['omega_2'] * omega['omega_2']
print(f"For A = {{omega_1, omega_2}}:")
print(f"LHS = {LHS_A_Y0:.4f}, RHS = {RHS_A_Y0:.4f}. Match: {np.isclose(LHS_A_Y0, RHS_A_Y0)}")

# Verify for A = {omega_3, omega_4} (Y=1)
# LHS
LHS_A_Y1 = EX_given_Y1 * omega['omega_3'] + EX_given_Y1 * omega['omega_4']
# RHS
RHS_A_Y1 = X_values['omega_3'] * omega['omega_3'] + X_values['omega_4'] * omega['omega_4']
print(f"\nFor A = {{omega_3, omega_4}}:")
print(f"LHS = {LHS_A_Y1:.4f}, RHS = {RHS_A_Y1:.4f}. Match: {np.isclose(LHS_A_Y1, RHS_A_Y1)}")

# Verify for A = Omega
# LHS = E[E[X|Y]]
LHS_A_Omega = EX_given_Y0 * P_Y0 + EX_given_Y1 * P_Y1
# RHS = E[X]
RHS_A_Omega = sum(X_values[k] * v for k, v in omega.items())
print(f"\nFor A = Omega:")
print(f"LHS = {LHS_A_Omega:.4f}, RHS = {RHS_A_Omega:.4f}. Match: {np.isclose(LHS_A_Omega, RHS_A_Omega)}")

For A = {omega_1, omega_2}:
LHS = 0.1000, RHS = 0.1000. Match: True

For A = {omega_3, omega_4}:
LHS = 0.2000, RHS = 0.2000. Match: True

For A = Omega:
LHS = 0.3000, RHS = 0.3000. Match: True


### 6. Find the Conditional PMF $p_{X|Y}(x \mid y)$.

The Conditional PMF $p_{X|Y}(x \mid y) = P(X=x \mid Y=y) = \frac{P(X=x, Y=y)}{P(Y=y)}$.

First, let's list the joint probabilities $P(X=x, Y=y)$:

*   $P(X=0, Y=0)$: This occurs at $\omega_1$. So, $P(X=0, Y=0) = P(\{\omega_1\}) = 0.50$.
*   $P(X=1, Y=0)$: This occurs at $\omega_2$. So, $P(X=1, Y=0) = P(\{\omega_2\}) = 0.10$.
*   $P(X=0, Y=1)$: This occurs at $\omega_3$. So, $P(X=0, Y=1) = P(\{\omega_3\}) = 0.20$.
*   $P(X=1, Y=1)$: This occurs at $\omega_4$. So, $P(X=1, Y=1) = P(\{\omega_4\}) = 0.20$.

Marginal probabilities for $Y$:
*   $P(Y=0) = P(\{\omega_1\}) + P(\{\omega_2\}) = 0.50 + 0.10 = 0.60$
*   $P(Y=1) = P(\{\omega_3\}) + P(\{\omega_4\}) = 0.20 + 0.20 = 0.40$

Now, we can compute the conditional PMF:

*   **Given $Y=0$:**
    *   $p_{X|Y}(0 \mid 0) = P(X=0 \mid Y=0) = \frac{P(X=0, Y=0)}{P(Y=0)} = \frac{0.50}{0.60} = \frac{5}{6}$
    *   $p_{X|Y}(1 \mid 0) = P(X=1 \mid Y=0) = \frac{P(X=1, Y=0)}{P(Y=0)} = \frac{0.10}{0.60} = \frac{1}{6}$

*   **Given $Y=1$:**
    *   $p_{X|Y}(0 \mid 1) = P(X=0 \mid Y=1) = \frac{P(X=0, Y=1)}{P(Y=1)} = \frac{0.20}{0.40} = \frac{2}{4} = \frac{1}{2}$
    *   $p_{X|Y}(1 \mid 1) = P(X=1 \mid Y=1) = \frac{P(X=1, Y=1)}{P(Y=1)} = \frac{0.20}{0.40} = \frac{2}{4} = \frac{1}{2}$

Summary of $p_{X|Y}(x \mid y)$:
*   $p_{X|Y}(0 \mid 0) = 5/6$
*   $p_{X|Y}(1 \mid 0) = 1/6$
*   $p_{X|Y}(0 \mid 1) = 1/2$
*   $p_{X|Y}(1 \mid 1) = 1/2$

In [5]:
# Joint Probabilities P(X=x, Y=y)
P_X0Y0 = omega['omega_1'] # X=0, Y=0 at omega_1
P_X1Y0 = omega['omega_2'] # X=1, Y=0 at omega_2
P_X0Y1 = omega['omega_3'] # X=0, Y=1 at omega_3
P_X1Y1 = omega['omega_4'] # X=1, Y=1 at omega_4

# Marginal Probabilities P(Y=y)
P_Y0 = P_X0Y0 + P_X1Y0 # 0.50 + 0.10 = 0.60
P_Y1 = P_X0Y1 + P_X1Y1 # 0.20 + 0.20 = 0.40

print("Conditional PMF p_X|Y(x | y):")
# For Y=0
p_X0_Y0 = P_X0Y0 / P_Y0
p_X1_Y0 = P_X1Y0 / P_Y0
print(f"  P(X=0 | Y=0) = {p_X0_Y0:.4f} (5/6)")
print(f"  P(X=1 | Y=0) = {p_X1_Y0:.4f} (1/6)")

# For Y=1
p_X0_Y1 = P_X0Y1 / P_Y1
p_X1_Y1 = P_X1Y1 / P_Y1
print(f"  P(X=0 | Y=1) = {p_X0_Y1:.4f} (1/2)")
print(f"  P(X=1 | Y=1) = {p_X1_Y1:.4f} (1/2)")

Conditional PMF p_X|Y(x | y):
  P(X=0 | Y=0) = 0.8333 (5/6)
  P(X=1 | Y=0) = 0.1667 (1/6)
  P(X=0 | Y=1) = 0.5000 (1/2)
  P(X=1 | Y=1) = 0.5000 (1/2)


### 7. Verify the identity: $$ \mathbb{E}[X\mid Y=y] = \sum_{x \in \{0,1\}} x \cdot p_{X|Y}(x \mid y) $$

We will verify this identity for $Y=0$ and $Y=1$.

*   **For $Y=0$:**
    $$ \mathbb{E}[X\mid Y=0] = (0) \cdot p_{X|Y}(0 \mid 0) + (1) \cdot p_{X|Y}(1 \mid 0) $$
    $$ = (0) \cdot \frac{5}{6} + (1) \cdot \frac{1}{6} = \frac{1}{6} $$
    This matches our earlier calculation for $E[X\mid Y=0]$.

*   **For $Y=1$:**
    $$ \mathbb{E}[X\mid Y=1] = (0) \cdot p_{X|Y}(0 \mid 1) + (1) \cdot p_{X|Y}(1 \mid 1) $$
    $$ = (0) \cdot \frac{1}{2} + (1) \cdot \frac{1}{2} = \frac{1}{2} $$
    This matches our earlier calculation for $E[X\mid Y=1]$.

The identity is verified.

In [6]:
# Verify for Y=0
EX_given_Y0_from_pmf = 0 * p_X0_Y0 + 1 * p_X1_Y0
print(f"E[X|Y=0] calculated from PMF = {EX_given_Y0_from_pmf:.4f}. Matches previous: {np.isclose(EX_given_Y0, EX_given_Y0_from_pmf)}")

# Verify for Y=1
EX_given_Y1_from_pmf = 0 * p_X0_Y1 + 1 * p_X1_Y1
print(f"E[X|Y=1] calculated from PMF = {EX_given_Y1_from_pmf:.4f}. Matches previous: {np.isclose(EX_given_Y1, EX_given_Y1_from_pmf)}")

E[X|Y=0] calculated from PMF = 0.1667. Matches previous: True
E[X|Y=1] calculated from PMF = 0.5000. Matches previous: True


### 8. Calculate the Mean Square Error (MSE)

The Mean Square Error (MSE) of $E[X\mid Y]$ as an estimator for $X$ is given by $E[(X - E[X\mid Y])^2]$.

We can calculate this by iterating through the outcomes $\omega_i$:

*   **For $\omega_1$ ($X=0, Y=0$):** $(X(\omega_1) - E[X\mid Y=0])^2 P(\{\omega_1\}) = (0 - 1/6)^2 \cdot 0.50 = (1/36) \cdot 0.50 = 0.50/36 = 1/72$
*   **For $\omega_2$ ($X=1, Y=0$):** $(X(\omega_2) - E[X\mid Y=0])^2 P(\{\omega_2\}) = (1 - 1/6)^2 \cdot 0.10 = (5/6)^2 \cdot 0.10 = (25/36) \cdot 0.10 = 2.5/36 = 1/14.4 = 5/72$
*   **For $\omega_3$ ($X=0, Y=1$):** $(X(\omega_3) - E[X\mid Y=1])^2 P(\{\omega_3\}) = (0 - 1/2)^2 \cdot 0.20 = (1/4) \cdot 0.20 = 0.05 = 1/20$
*   **For $\omega_4$ ($X=1, Y=1$):** $(X(\omega_4) - E[X\mid Y=1])^2 P(\{\omega_4\}) = (1 - 1/2)^2 \cdot 0.20 = (1/4) \cdot 0.20 = 0.05 = 1/20$

Summing these values:
$$ MSE = \frac{1}{72} + \frac{5}{72} + \frac{1}{20} + \frac{1}{20} = \frac{6}{72} + \frac{2}{20} = \frac{1}{12} + \frac{1}{10} = \frac{5+6}{60} = \frac{11}{60} $$

So, $MSE = \frac{11}{60} \approx 0.1833$.

In [7]:
# Calculate MSE = E[(X - E[X|Y])^2]
mse = 0

mse += (X_values['omega_1'] - EX_given_Y0)**2 * omega['omega_1'] # (0 - 1/6)^2 * 0.50
mse += (X_values['omega_2'] - EX_given_Y0)**2 * omega['omega_2'] # (1 - 1/6)^2 * 0.10
mse += (X_values['omega_3'] - EX_given_Y1)**2 * omega['omega_3'] # (0 - 1/2)^2 * 0.20
mse += (X_values['omega_4'] - EX_given_Y1)**2 * omega['omega_4'] # (1 - 1/2)^2 * 0.20

print(f"Mean Square Error (MSE) = {mse:.4f} (which is 11/60)")

Mean Square Error (MSE) = 0.1833 (which is 11/60)


### 9. How would the Mean Square Error (MSE) change if the bank decided to ignore the customer risk class $Y$ and instead used the global average fraud rate ($E[X] = 0.30$) as a constant estimate for every transaction? Calculate this error and explain its relationship to the Law of Total Variance.

If the bank ignores $Y$ and uses the global average fraud rate $E[X]$ as an estimate, the estimator is simply $E[X]$ (a constant). The MSE in this case is $E[(X - E[X])^2]$, which is the variance of $X$, $Var(X)$.

First, let's calculate $E[X]$:
$$ E[X] = (0)(0.50) + (1)(0.10) + (0)(0.20) + (1)(0.20) = 0.10 + 0.20 = 0.30 $$
This matches the given global average fraud rate.

Now, let's calculate $E[X^2]$:
$$ E[X^2] = (0^2)(0.50) + (1^2)(0.10) + (0^2)(0.20) + (1^2)(0.20) = 0.10 + 0.20 = 0.30 $$

The variance of $X$ is:
$$ Var(X) = E[X^2] - (E[X])^2 = 0.30 - (0.30)^2 = 0.30 - 0.09 = 0.21 $$

So, if the bank ignores $Y$, the MSE would be $0.21$. This is higher than the MSE when using $Y$ ($0.1833$), which makes sense because using more information (the customer risk class) should lead to a better (lower MSE) estimator.

**Relationship to the Law of Total Variance:**
The Law of Total Variance states:
$$ Var(X) = E[Var(X\mid Y)] + Var(E[X\mid Y]) $$

We know that $E[Var(X\mid Y)]$ is the MSE of $E[X\mid Y]$ as an estimator for $X$, i.e., $E[(X - E[X\mid Y])^2]$.
Let's calculate $Var(E[X\mid Y])$.

$E[X\mid Y]$ takes values $1/6$ with probability $P(Y=0) = 0.60$ and $1/2$ with probability $P(Y=1) = 0.40$.

*   $E[E[X\mid Y]] = (1/6)(0.60) + (1/2)(0.40) = 0.10 + 0.20 = 0.30$ (This is $E[X]$ as expected by the Law of Total Expectation).

*   $E[(E[X\mid Y])^2] = (1/6)^2 (0.60) + (1/2)^2 (0.40)$
    $$ = (1/36)(0.60) + (1/4)(0.40) = 0.60/36 + 0.10 = 1/60 + 1/10 = 1/60 + 6/60 = 7/60 $$

*   $Var(E[X\mid Y]) = E[(E[X\mid Y])^2] - (E[E[X\mid Y]])^2 = 7/60 - (0.30)^2 = 7/60 - 0.09 = 7/60 - 9/100$
    $$ = \frac{35 - 27}{300} = \frac{8}{300} = \frac{2}{75} \approx 0.02666 $$

Now, let's verify the Law of Total Variance:
$$ Var(X) = E[(X - E[X\mid Y])^2] + Var(E[X\mid Y]) $$
$$ 0.21 = \frac{11}{60} + \frac{2}{75} $$
$$ 0.21 = 0.18333... + 0.02666... $$
$$ 0.21 = 0.21000... $$

The relationship is that the total variance $Var(X)$ is decomposed into two parts: the variance *within* the conditional groups ($E[Var(X\mid Y)]$, which is our MSE), and the variance *between* the conditional means ($Var(E[X\mid Y)]$). Using $E[X]$ as an estimator essentially means that the information from $Y$ is not used, leading to an MSE equal to $Var(X)$, which is always greater than or equal to $E[Var(X|Y)]$. The difference $Var(E[X|Y])$ represents the reduction in MSE achieved by conditioning on $Y$.

In [8]:
# Calculate E[X]
EX = sum(X_values[k] * v for k, v in omega.items())
print(f"E[X] = {EX:.4f}")

# Calculate MSE if using E[X] as estimator (Var(X))
EX2 = sum(X_values[k]**2 * v for k, v in omega.items())
VarX = EX2 - EX**2
print(f"MSE if ignoring Y (Var(X)) = {VarX:.4f}")

# Calculate Var(E[X|Y]) for Law of Total Variance
# E[X|Y] takes values EX_given_Y0 with P_Y0, and EX_given_Y1 with P_Y1
E_EX_given_Y = EX_given_Y0 * P_Y0 + EX_given_Y1 * P_Y1 # Should be equal to EX
E_EX_given_Y_squared = (EX_given_Y0**2 * P_Y0) + (EX_given_Y1**2 * P_Y1)
Var_EX_given_Y = E_EX_given_Y_squared - E_EX_given_Y**2
print(f"Var(E[X|Y]) = {Var_EX_given_Y:.4f} (which is 2/75)")

# Verify Law of Total Variance: Var(X) = E[Var(X|Y)] + Var(E[X|Y])
# E[Var(X|Y)] is our calculated MSE_conditional
print(f"\nVerifying Law of Total Variance:")
print(f"Var(X) = {VarX:.4f}")
print(f"E[Var(X|Y)] + Var(E[X|Y]) = {mse:.4f} + {Var_EX_given_Y:.4f} = {mse + Var_EX_given_Y:.4f}")
print(f"Match: {np.isclose(VarX, mse + Var_EX_given_Y)}")

E[X] = 0.3000
MSE if ignoring Y (Var(X)) = 0.2100
Var(E[X|Y]) = 0.0267 (which is 2/75)

Verifying Law of Total Variance:
Var(X) = 0.2100
E[Var(X|Y)] + Var(E[X|Y]) = 0.1833 + 0.0267 = 0.2100
Match: True


### 10. Using the joint distribution of $(X, Y)$ derived from the sample space:

Let's first define the joint probability distribution $P(X=x, Y=y)$.

| $X \setminus Y$ | $Y=0$ (low-risk) | $Y=1$ (high-risk) | $P(X=x)$ |
|:---------------:|:----------------:|:-----------------:|:----------:|
| $X=0$ (legit)   | $P(\omega_1)=0.50$ | $P(\omega_3)=0.20$ | $0.70$     |
| $X=1$ (fraud)   | $P(\omega_2)=0.10$ | $P(\omega_4)=0.20$ | $0.30$     |
| $P(Y=y)$        | $0.60$           | $0.40$            | $1.00$     |

From this table:
*   $P(X=0) = 0.50 + 0.20 = 0.70$
*   $P(X=1) = 0.10 + 0.20 = 0.30$
*   $P(Y=0) = 0.50 + 0.10 = 0.60$
*   $P(Y=1) = 0.20 + 0.20 = 0.40$

In [9]:
from scipy.stats import entropy

# Marginal probabilities for X
P_X0 = P_X0Y0 + P_X0Y1 # 0.50 + 0.20 = 0.70
P_X1 = P_X1Y0 + P_X1Y1 # 0.10 + 0.20 = 0.30
P_X = [P_X0, P_X1]

# Marginal probabilities for Y (already calculated)
P_Y = [P_Y0, P_Y1]

print(f"P(X=0) = {P_X0:.2f}, P(X=1) = {P_X1:.2f}")
print(f"P(Y=0) = {P_Y0:.2f}, P(Y=1) = {P_Y1:.2f}")

P(X=0) = 0.70, P(X=1) = 0.30
P(Y=0) = 0.60, P(Y=1) = 0.40


#### a) Compute the entropy $H(X)$, representing the bank's initial uncertainty regarding fraud.

The entropy $H(X)$ is calculated as:
$$ H(X) = - \sum_{x} P(X=x) \log_2 P(X=x) $$
$$ H(X) = - [P(X=0) \log_2 P(X=0) + P(X=1) \log_2 P(X=1)] $$
$$ H(X) = - [0.70 \log_2(0.70) + 0.30 \log_2(0.30)] $$
$$ H(X) \approx - [0.70 \cdot (-0.5146) + 0.30 \cdot (-1.7370)] $$
$$ H(X) \approx - [-0.3602 - 0.5211] \approx 0.8813 \text{ bits} $$

In [10]:
# H(X)
H_X = entropy(P_X, base=2)
print(f"H(X) = {H_X:.4f} bits")

H(X) = 0.8813 bits


#### b) Compute the conditional entropy $H(X \mid Y)$, representing the uncertainty remaining after the customer's risk class is revealed.

The conditional entropy $H(X \mid Y)$ is calculated as:
$$ H(X \mid Y) = - \sum_{y} P(Y=y) \sum_{x} p_{X|Y}(x \mid y) \log_2 p_{X|Y}(x \mid y) $$
$$ H(X \mid Y) = P(Y=0) H(X \mid Y=0) + P(Y=1) H(X \mid Y=1) $$

Where:
*   $H(X \mid Y=0) = - [p_{X|Y}(0 \mid 0) \log_2 p_{X|Y}(0 \mid 0) + p_{X|Y}(1 \mid 0) \log_2 p_{X|Y}(1 \mid 0)]$
    $$ = - [\frac{5}{6} \log_2(\frac{5}{6}) + \frac{1}{6} \log_2(\frac{1}{6})] $$
    $$ \approx - [0.8333 \cdot (-0.2630) + 0.1667 \cdot (-2.5850)] $$
    $$ \approx - [-0.2191 - 0.4309] \approx 0.6500 \text{ bits} $$

*   $H(X \mid Y=1) = - [p_{X|Y}(0 \mid 1) \log_2 p_{X|Y}(0 \mid 1) + p_{X|Y}(1 \mid 1) \log_2 p_{X|Y}(1 \mid 1)]$
    $$ = - [\frac{1}{2} \log_2(\frac{1}{2}) + \frac{1}{2} \log_2(\frac{1}{2})] = - [0.5 \cdot (-1) + 0.5 \cdot (-1)] $$
    $$ = - [-0.5 - 0.5] = 1 \text{ bit} $$

Now, combine these:
$$ H(X \mid Y) = (0.60) \cdot H(X \mid Y=0) + (0.40) \cdot H(X \mid Y=1) $$
$$ H(X \mid Y) \approx (0.60) \cdot (0.6500) + (0.40) \cdot (1) $$
$$ H(X \mid Y) \approx 0.3900 + 0.40 = 0.7900 \text{ bits} $$

In [11]:
# Conditional entropy H(X|Y)

# H(X|Y=0)
P_X_given_Y0 = [p_X0_Y0, p_X1_Y0]
H_X_given_Y0 = entropy(P_X_given_Y0, base=2)
print(f"H(X|Y=0) = {H_X_given_Y0:.4f} bits")

# H(X|Y=1)
P_X_given_Y1 = [p_X0_Y1, p_X1_Y1]
H_X_given_Y1 = entropy(P_X_given_Y1, base=2)
print(f"H(X|Y=1) = {H_X_given_Y1:.4f} bits")

H_X_given_Y = P_Y0 * H_X_given_Y0 + P_Y1 * H_X_given_Y1
print(f"\nH(X|Y) = {H_X_given_Y:.4f} bits")

H(X|Y=0) = 0.6500 bits
H(X|Y=1) = 1.0000 bits

H(X|Y) = 0.7900 bits


#### c) Calculate the Mutual Information $I(X; Y)$.

The Mutual Information $I(X; Y)$ can be calculated as:
$$ I(X; Y) = H(X) - H(X \mid Y) $$
Using the values from above:
$$ I(X; Y) \approx 0.8813 - 0.7900 = 0.0913 \text{ bits} $$

In [12]:
# Mutual Information I(X;Y)
I_XY = H_X - H_X_given_Y
print(f"I(X;Y) = {I_XY:.4f} bits")

I(X;Y) = 0.0913 bits


#### d) Based on these values, what percentage of the total uncertainty about fraud is removed by knowing the customer's risk class?

The percentage of total uncertainty about fraud ($H(X)$) that is removed by knowing the customer's risk class ($Y$) is given by the ratio of Mutual Information to the initial entropy of $X$, multiplied by 100.

$$ \text{Percentage Removed} = \frac{I(X; Y)}{H(X)} \times 100\% $$
$$ \text{Percentage Removed} \approx \frac{0.0913}{0.8813} \times 100\% \approx 10.36\% $$

In [13]:
# Percentage of uncertainty removed
percentage_removed = (I_XY / H_X) * 100
print(f"Percentage of total uncertainty removed by knowing Y: {percentage_removed:.2f}%")

Percentage of total uncertainty removed by knowing Y: 10.36%
